homeassistant/components/nederlandse_spoorwegen/config_flow.py

```diff
@@ -0,0 +1,141 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.api_key = user_input[CONF_API_KEY]
+            try:
+                await get_stations(ns_api.NSAPI(self.api_key))
+            except RequestParametersError:
+                errors["base"] = "invalid_api_key"
+            else:
+                await self.async_set_unique_id(self.api_key)
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title="NS API", data={CONF_API_KEY: self.api_key}
+                )
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema({vol.Required(CONF_API_KEY): str}),
+            errors=errors,
+        )
+
+
+class NederlandseSpoorwegenSubentryFlowHandler(ConfigSubentryFlow):
+    """Handle subentry flow for adding and modifying a location."""
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._stations: dict[str, str] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """User flow to add a new train."""
+
+        if len(self._stations) == 0:
+            nsapi = self.hass.config_entries.async_get_known_entry(
+                self.handler[0]
+            ).runtime_data.nsapi
+
+            self._stations = {
+                station.code: station.names["middle"]
+                for station in await get_stations(nsapi)
+                if station.country == "NL"
+            }
+
+            self._stations = dict(
+                sorted(self._stations.items(), key=lambda item: item[1])
+            )
+
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            title = user_input.get(CONF_NAME, None)
+
+            if title is None:
+                title = self._stations[user_input[CONF_STATION_FROM]]
+                if CONF_STATION_VIA in user_input:
+                    title = title + " - " + self._stations[user_input[CONF_STATION_VIA]]
+                title = title + " - " + self._stations[user_input[CONF_STATION_TO]]
+                if CONF_TIME in user_input:
+                    title = title + " - " + user_input[CONF_TIME]
+
+            return self.async_create_entry(
+                title=title,
+                data={
+                    CONF_STATION_FROM: user_input[CONF_STATION_FROM],
+                    CONF_STATION_VIA: user_input.get(CONF_STATION_VIA, None),
+                    CONF_STATION_TO: user_input[CONF_STATION_TO],
+                    CONF_TIME: user_input.get(CONF_TIME, None),
+                },
+            )
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(CONF_NAME): str,
```

Comment:
I think we should allow the user to rename the config entry via HA. I don't believe that is supported right now, but that would be ideal.